src/python/dxpy/__init__.py

```diff
@@ -288,13 +288,15 @@ def DXHTTPRequest(resource, data, method='POST', headers=None, auth=True, timeou
         rewind_input_buffer_offset = data.tell()
 
     last_exc_type, last_error, last_traceback = None, None, None
+    time_started = time.time() if timeout else None
     try_index = 0
     while True:
         success, streaming_response_truncated = True, False
         response = None
         try:
             _method, _url, _headers = _process_method_url_headers(method, url, headers)
-            response = session_handler.request(_method, _url, headers=_headers, data=data, timeout=timeout, auth=auth,
+            _timeout = timeout or 600
```

Comment:
IMO it's better to preserve the simpler semantics of passing the unmodified timeout value to requests (which then sets it on the socket) than it is to change it to ensure that we spend no more than `timeout` before returning (which is not a guarantee that the caller can fully rely on anyway, since it's not anything like a real-time system). The scenario of the first few requests returning quickly and the last one hanging (which is necessary to trigger the scenario of significantly missing the timeout deadline before returning) seems far-fetched to me. I agree there is a minor correctness issue, but the two applications of the timeout value really deal with different issues: the one passed to requests and set on the socket takes care of stalled out requests, while the one checked in the retry loop takes care of breaking out of the 503 retry-later mode.

(We may also want to investigate using urllib3's retry support, which was introduced over the course of 2014. Maybe it better deals with this socket/total timeout duality?)